incompatibility/incompatibility-report/oas3-traverse.go

```diff
@@ -0,0 +1,160 @@
+}
+
+func (e TraverseError) Error() string {
+	return e.desrciptorString
+}
+
+// Structure for generic single-input single-output function with error indication
+type GenericOperation struct {
+	op         func(interface{}) (interface{}, error)
+	inputType  reflect.Type
+	outputType reflect.Type
+}
+
+type transverseComp int
+
+const (
+	DiffType = iota
+	SameType
+)
+
+func (op1 GenericOperation) typeEquality(op2 GenericOperation) transverseComp {
+	if op1.inputType == op2.inputType && op1.outputType == op2.outputType {
+		return SameType
+	}
+	return DiffType
+}
+
+// arg function should be a single-input single-output function
+// creates generic wrapper around function, indicates error on return
+func makeGenericOperation(function interface{}) (GenericOperation, error) {
+	switch inputType := reflect.TypeOf(function); inputType.Kind() {
+	case reflect.Func:
+		// Continue function
+		validFunction := inputType.NumIn() == 1 && inputType.NumOut() == 1
+		if validFunction {
+			funcInputT := inputType.In(0)
+			funcOutputT := inputType.Out(0)
+			formattedGenericFunction := func(input interface{}) (interface{}, error) {
+				switch t := reflect.TypeOf(input); t {
+				case funcInputT:
+					functionAsValue := reflect.ValueOf(function)
+					returnAsInterface := functionAsValue.Call([]reflect.Value{reflect.ValueOf(input)})[0].Interface()
+					returnType := reflect.TypeOf(returnAsInterface)
+					if returnType != funcOutputT {
+						return nil, TraverseError{"Bad Return Type in Invocation"}
+					}
+					return returnAsInterface, nil
+				default:
+					return nil, TraverseError{"Bad Input Type in Inovocation"}
+				}
+			}
+			return GenericOperation{
+				op:         formattedGenericFunction,
+				inputType:  funcInputT,
+				outputType: funcOutputT,
+			}, nil
+		}
+		goto failGen
+	default:
+		// Invalid input
+		goto failGen
+	}
+failGen:
+	return GenericOperation{}, TraverseError{"input is not a single-input single-output function"}
+}
+
+// function should be a single-input single-output function whose output type is *IncompatibilityReport
+// performs error checking for these requirements and casts into generic type
+func makeGenericIncompatibilityReportFunc(function interface{}) (func(interface{}) (*IncompatibilityReport, error), error) {
+	genGeneric, err := makeGenericOperation(function)
+	badType := genGeneric.outputType != reflect.TypeOf(&IncompatibilityReport{})
+	if err != nil || badType {
+		return nil, err
+	} else {
+		return func(i interface{}) (*IncompatibilityReport, error) {
+			res, err := genGeneric.op(i)
+			return res.(*IncompatibilityReport), err
+		}, nil
+	}
+}
+
+// ComponentType - component type in OpenAPIDoc, i.e. Parameters, Components, ...
+// ParentTraverse - lambda to tranverse from parent to component of type ComponentType
+// OperationOnComponent - lambda which looks at current layer for incompatibilities
+// ChildPaths - paths to check for deeper incompatibilities
+type PathOperation struct {
```

Comment:
This struct doesn't seem like it belongs in this package. This has more to do with generic operations on a OpenAPI spec, rather than specific incompatibility checks. Consider refactoring this class into another package.

Reply:
More on this class:
I read it as more of a `OpenAPIComponentOperation`? It seems like it is suppose to encapsulate what `ComponentType` it works on, how to get there in the spec, and an operation on that component. 

Is your intent that these operations can be applied at any level in the spec?